addons/procurement/procurement.py

```diff
@@ -238,59 +196,40 @@ def _find_suitable_rule(self, cr, uid, procurement, context=None):
         '''
         return False
 
-    def _assign(self, cr, uid, procurement, context=None):
-        '''This method check what to do with the given procurement in order to complete its needs.
-        It returns False if no solution is found, otherwise it stores the matching rule (if any) and
-        returns True.
-            :param procurement: browse record
-            :rtype: boolean
-        '''
+    @api.model
+    def _assign(self, procurement):
         #if the procurement already has a rule assigned, we keep it (it has a higher priority as it may have been chosen manually)
         if procurement.rule_id:
             return True
         elif procurement.product_id.type != 'service':
-            rule_id = self._find_suitable_rule(cr, uid, procurement, context=context)
+            rule_id = self._find_suitable_rule(procurement)
             if rule_id:
-                self.write(cr, uid, [procurement.id], {'rule_id': rule_id}, context=context)
+                procurement.write({'rule_id': rule_id})
                 return True
         return False
 
-    def _run(self, cr, uid, procurement, context=None):
+    @api.model
+    def _run(self, procurement):
```

Comment:
Here @api.model method is not appropriate. You can make a method for @api.v7 and @api.v8 .